### PR TITLE
Add L1HandlerTransactionTrace

### PR DESCRIPTION
**Pull request by mikiw**
## Usage related changes

- Add support of `L1_HANDLER` transaction traces in `starknet_traceTransaction` `starknet_traceBlockTransactions` RPC methods

## Development related changes

- Although `json_rpc_client.trace_transaction()` method in integration tests in `test_trace.rs` is working properly, currently it doesn't work for `L1_HANDLER` transaction traces and I needed to use custom rpc call

## Checklist:

- [x] Applied formatting - `./scripts/format.sh`
- [x] No linter errors - `./scripts/clippy_check.sh`
- [x] No unused dependencies - `./scripts/check_unused_deps.sh`
- [x] Performed code self-review
- [x] Rebased to the last commit of the target branch (or merged it into my branch)
- [x] Documented the changes
- [x] Linked the issues which this PR resolves
- [x] Checked the TODO section in README.md if this PR resolves it
- [x] Updated the tests
- [x] All tests are passing - `cargo test`

### CODE REVIEW

**Comment by mikiw** [comment]
I'm not sure why `function_invocation` is null, I'm still checking that.

File: crates/starknet-server/tests/test_messaging.rs [hunk 1]
```diff
@@ -481,6 +481,19 @@ mod test_messaging {
             devnet.post_json("/postman/flush".into(), "".into()).await.expect("flush failed");
         assert_eq!(resp.status(), StatusCode::OK, "Checking status of {resp:?}");
 
+        // Assert traces of L1Handler transaction with custom rpc call,
```
Comment (FabijanC) [inline crates/starknet-server/tests/test_messaging.rs:484]: We already have a pretty long test case here, adding this is only making it longer. Could this addition be a separate test, like `assert_l1_handler_tx_can_be_dumped_and_loaded`? If not, at least L487-L492 could be extracted into a separate function for a slight improvement.

File: crates/starknet-server/tests/test_messaging.rs [hunk 2]
```diff
@@ -481,6 +481,19 @@ mod test_messaging {
...
+                "starknet_traceTransaction",
+                json!({ "transaction_hash": flush_body.get("generated_l2_transactions").unwrap()[0] }),
+            )
+            .await["result"];
+        assert_eq!(l1_handler_tx_trace["type"], "L1_HANDLER");
+        assert!(l1_handler_tx_trace["function_invocation"].is_null());
+        assert!(l1_handler_tx_trace["state_diff"].is_null());
+
         // Ensure the balance is back to 1 on L2.
         assert_eq!(get_balance(&devnet, sn_l1l2_contract, user_sn).await, [FieldElement::ONE]);
```
Comment (FabijanC) [inline crates/starknet-server/tests/test_messaging.rs:498]: In any case, this assertion should be done sooner, just after flush response status is checked in L482.

**Review by FabijanC — CHANGES_REQUESTED** [review]

File: crates/starknet-server/tests/test_messaging.rs [hunk 3]
```diff
@@ -481,6 +481,19 @@ mod test_messaging {
...
+        let flush_body = get_json_body(resp).await;
+        let l1_handler_tx_trace = &devnet
+            .send_custom_rpc(
+                "starknet_traceTransaction",
+                json!({ "transaction_hash": flush_body.get("generated_l2_transactions").unwrap()[0] }),
+            )
+            .await["result"];
+        assert_eq!(l1_handler_tx_trace["type"], "L1_HANDLER");
+        assert!(l1_handler_tx_trace["function_invocation"].is_null());
+        assert!(l1_handler_tx_trace["state_diff"].is_null());
```
[the remark this answers is not in the document]
Comment (mikiw) [reply crates/starknet-server/tests/test_messaging.rs]: I think it's not a good idea, these objects can be huge

File: crates/types/src/rpc/transactions.rs [hunk 4]
```diff
@@ -718,6 +720,13 @@ pub struct DeployAccountTransactionTrace {
     pub state_diff: Option<ThinStateDiff>,
 }
 
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(deny_unknown_fields)]
+pub struct L1HandlerTransactionTrace {
+    pub function_invocation: Option<FunctionInvocation>,
```
[the remark this answers is not in the document]
Comment (mikiw) [reply crates/types/src/rpc/transactions.rs]: Fixed in 4c893e5fce23c55c1188b89b9f211b269991d910 and 7d96523719246b758ed0e0da87aac97079b372cc

File: crates/starknet-server/tests/test_messaging.rs (hunk 2 above)
Comment (mikiw) [reply crates/starknet-server/tests/test_messaging.rs]: ed5595b7bfbce2330aa5cfa29532148f735ddb07

File: crates/starknet-server/tests/test_messaging.rs (hunk 1 above)
Comment (mikiw) [reply crates/starknet-server/tests/test_messaging.rs:501]: 4a15d01c319ef3ca8c2321a44b705767cf584cda

File: crates/starknet-server/tests/test_messaging.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (FabijanC) [reply crates/starknet-server/tests/test_messaging.rs]: the comment was written with a smaller `function_invocation` in mind. Let's then keep it as it is. But also let's keep in mind that comparing complete objects can be useful.

**Review by FabijanC — APPROVED** [review]

**Pull request merged**